File: Dock/Globals/Classes/Credits/CreditRewardMilestone.js
// A one-time reward milestone: once a user's cumulative lifetime spend
// reaches `spendThreshold`, they are granted `rewardCredits` exactly once
// (idempotency is enforced by CreditLedger via the reward reference key).

class CreditRewardMilestone
{
    #spendThreshold;
    #rewardCredits;

    constructor({ spendThreshold = 0, rewardCredits = 0 } = {})
    {
        this.setSpendThreshold(spendThreshold);
        this.setRewardCredits(rewardCredits);
    }

    getSpendThreshold()
    {
        return this.#spendThreshold;
    }

    setSpendThreshold(value)
    {
        value = parseFloat(value);
        if (isNaN(value) || value < 0)
        {
            value = 0;
        }
        this.#spendThreshold = value;
    }

    getRewardCredits()
    {
        return this.#rewardCredits;
    }

    setRewardCredits(value)
    {
        value = parseFloat(value);
        if (isNaN(value) || value < 0)
        {
            value = 0;
        }
        this.#rewardCredits = value;
    }

    toJson()
    {
        return {
            spendThreshold: this.getSpendThreshold(),
            rewardCredits: this.getRewardCredits(),
        };
    }

    static fromJson(json)
    {
        return new CreditRewardMilestone({
            spendThreshold: json?.spendThreshold ?? 0,
            rewardCredits: json?.rewardCredits ?? 0,
        });
    }
}

module.exports = CreditRewardMilestone;

File: Agent/Globals/Classes/Credits/CreditRewardMilestone.py
# Mirror of Dock/Globals/Classes/Credits/CreditRewardMilestone.js.


class CreditRewardMilestone:

    def __init__(self, spend_threshold: float = 0, reward_credits: float = 0) -> None:
        self.set_spend_threshold(spend_threshold)
        self.set_reward_credits(reward_credits)

    def get_spend_threshold(self) -> float:
        return self.__spend_threshold

    def set_spend_threshold(self, value) -> None:
        try:
            value = float(value)
        except (TypeError, ValueError):
            value = 0
        if value < 0:
            value = 0
        self.__spend_threshold = value

    def get_reward_credits(self) -> float:
        return self.__reward_credits

    def set_reward_credits(self, value) -> None:
        try:
            value = float(value)
        except (TypeError, ValueError):
            value = 0
        if value < 0:
            value = 0
        self.__reward_credits = value

    def to_json(self) -> dict:
        return {
            "spendThreshold": self.get_spend_threshold(),
            "rewardCredits": self.get_reward_credits(),
        }

    @classmethod
    def from_json(cls, data: dict) -> "CreditRewardMilestone":
        data = data or {}
        return cls(
            spend_threshold=data.get("spendThreshold", 0),
            reward_credits=data.get("rewardCredits", 0),
        )
